### DISCUSSION TITLE
Thermostat fails after TimeRampupMaxSet minutes

### DISCUSSION THREAD

**Discussion by wchulze** [arendst/Tasmota#16289] [Q&A]
I have a setup of a Wemos D1 Mini Pro with Tasmota V 11.1.0 acting as a heating controller with 1 relays.

The system runs independent with a local mounted DS18B20 temperature sensor.
MQTT interface is disabled and the HTTP API is enabled.
The Power supply can source > 1A at 5V.

The initial configuration is set by:

**Backlog Rule1 1; Sleep 40; SetOption0 0**	

and the following rule:

**Rule1 ON Power1#boot DO Backlog
SensorInputSet 1;
OutputRelaySet 1;
ThermostatModeSet 1;
ControllerModeSet 2;
TempTargetSet 29;
TimeRampupCycleSet 1;
TimePiCycleSet 1;
ENDON**

**Problem description:**
The controller works with these settings as expected until the preset T**imeRampupMaxSet = 960** has expired. Then controller **_always stops controling_** and sets the output into the OFF state.

I read along the issue _thermostat go to sleep after no heater power #13512_. No real solution found.

I study the documentation very carefully and find in the commands description of Tasmota for the thermostat:

**TimeRampupMaxSet  Sets the maximum time in minutes for the controller to stay in the ramp-up phase ...**

Does that really mean the total runtime for the controller in the ramp-up mode is limited to **TimeRampupMaxSet = 1440** ?
I examined the source code in the xdrv_39_thermostat.ino file.

The variable **Thermostat[ctr_output].timestamp_rampup_start**,
containig the value **TasmotaGlobal.uptime**, is set/chage within
Line 214 init struct at startup
Line 477 in function **ThermostatHybridCtrPhase** (not executed in ramp-up mode)
Line 1466 in function **CmndControllerModeSet**

How is the time controlled
Line 845: Calculate the running time
    `time_in_rampup = TasmotaGlobal.uptime - Thermostat[ctr_output].timestamp_rampup_start;`

Line 856: Checking if the time is done 
   `if ((time_in_rampup <= (60 * (uint32_t)Thermostat[ctr_output].time_rampup_max))...`

In fact, the value of **TimeRampupMaxSet** is the total run time of the controller for ramp-up operation. From my point of view we have the following problem after the time has expired:

The controller stops in predictive ramp-up mode. The output stay always at OFF. **This is the reported problem**.
Only a restart or **ControllerModeSet** reactivate the controller.

The variable **Thermostat[ctr_output].timestamp_rampup_start** may be set to **TasmotaGlobal.uptime** if the output of the controller is set to ON.

Assuming the heating system is designed for a minimum outside temperature of -15°C. If the outside temperature falls below this value, the desired room temperature (e.g. 20°C) can no longer be reached, but the heating remains active and keeps the room frost-free down to -45°C outside temperature. But after **TimeRampupMaxSet** minutes, the controller switches the output to OFF. Now the room temperature drops to the outside temperature!

The switch-off behavior of the controller after **TimeRampupMaxSet** minutes may be adjustable. **TimeRampupMaxSet = 0** could be used for this. I'm unsure about side effects.

Wolfgang

**Comment by dexterbot80** [comment]
Hello, I have the same problem
My thermostat stop working after ??? minutes.
Did you manage to solve the problem?
Thanks

**Comment by wchulze** [comment]
Hello

unfortunately I could not finally solve the problem. I have a problem report
Thermostat fails after 16 hours #16424
posted to https://github.com/arendst/Tasmota/issues. Please post your question there to keep the issue alive. You will find an exact error description as well as an approach to solve the problem in the software. So far I haven't received any feedback. Theo asked the author of the thermostat to take a look inside the software, without any reaction.

As an interim solution I have the simple thermostat
https://tasmota.github.io/docs/Rules/#simple-thermostat-example
implemented with some modifications. There is - from me - a Windows application to control the thermostat.

Wolfgang

**Reply by joba-1** [reply]
Seems to me that what you experience is an important safety feature.

Imagine what happens if the temperature sensor fails and constantly reports a temperature below setpoint (not unheard of…):

Even if the actual temperature is above healthy limits, it would not stop heating.

**Comment by wchulze** [comment]
Hi joba-1,
thanks for the comment. This is what I wrote:

**_The switch-off behavior of the controller after reaching TimeRampupMaxSet minutes should be adjustable._**

A defective sensor usually fails altogether. This should be recognised, e.g. that the temperature does not change over a defined period of time.It must also be recognized that the sensor is missing (Sernse Line Break). In this case the power must be set to  **Off**

Wolfgang
